pkg/cloud/services/gc/cleanup.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	rgapi "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/annotations"
+)
+
+const (
+	serviceNameTag    = "kubernetes.io/service-name"
+	eksClusterNameTag = "aws:eks:cluster-name"
+)
+
+// ReconcileDelete is responsible for determining if the infra cluster needs to be garbage collected. If
+// does then it will perform garbage collection. For example, it will delete the ELB/NLBs that where created
+// as a result of Services of type load balancer.
+func (s *Service) ReconcileDelete(ctx context.Context) error {
+	s.scope.Info("reconciling deletion for garbage collection")
+
+	val, found := annotations.Get(s.scope.InfraCluster(), expinfrav1.ExternalResourceGCAnnotation)
+	if !found {
+		val = "true"
+	}
+
+	shouldGC, err := strconv.ParseBool(val)
+	if err != nil {
+		return fmt.Errorf("converting value %s of annotation %s to bool: %w", val, expinfrav1.ExternalResourceGCAnnotation, err)
+	}
+
+	if shouldGC {
+		if err := s.deleteResources(ctx); err != nil {
+			return fmt.Errorf("deleting workload services of type load balancer: %w", err)
+		}
+	}
```

Comment:
nit: Return early instead of nesting a conditional.

```suggestion
	if !shouldGC {
		return nil
	}
	
	if err := s.deleteResources(ctx); err != nil {
		return fmt.Errorf("deleting workload services of type load balancer: %w", err)
	}
```

Reply:
Changed.